Create markdown document for CVals (#250)
youtube/cobalt: cobalt/doc/cvals.md

# Cobalt CVals

## Overview

CVals are globally accessible, thread-safe, key-value pairs within Cobalt, which
are primarily used for monitoring state and tracking performance and memory.
Each CVal is defined with a unique key and registered with the global
CValManager. The CValManager can subsequently be queried for the current value
of the key, which is returned as a string. CVals come in two varieties:

> PublicCVals, which are active in all builds; and DebugCVals, which are only

Reviewer: Consider changing to list.

```suggestion
CVals come in two varieties:

- PublicCVals, which are active in all builds
- DebugCVals, which are only enabled in non-Gold builds
```

enabled in non-Gold builds.

## Usage

> There are two ways to access CVals: through the Debug Console and via

Reviewer: Same, consider changing to list.

pressing Ctrl-O. It displays the current values of registered CVals, updating
them at 60 Hz. The initially displayed values are determined by
`DEFAULT_ACTIVE_SET` in `debug_values/console_values.js`. However, this is
modifiable in the debug console hud.

The registered CVals can be viewed by entering `debug.cvalList()` into the debug
console. Additional CVals can be registered via
`debug.cvalAdd(substringToMatch)` and can be removed via
`debug.cvalRemove(substringToMatch)`. The current registered list can be saved
as a new set via `debug.cvalSave(set_key)`. At that point, the saved set can
later be loaded as the active set via the command, `debug.cvalLoad(set_key)`.

A full list of the commands available via the debug console is shown by entering
`debug.help()`.

### JavaScript

CVals are exposed to JavaScript via the H5vccCVal object. This provides an
interface for requesting all of the keys, which are returned as an array, and
for retrieving a specific CVal value, which is returned as an optional string,
via its CVal key. While the H5vccCVal object is usable in any build, only public
CVals are queryable in Gold builds.

> Here are examples its usage:

Reviewer: ```suggestion
Here are examples of its usage:
```

  > h5vcc.cVal.getValue("Count.MainWebModule.Layout.Box")
  < "463"
```

## Tracking Categories

### Cobalt

#### PublicCVals

*   **Cobalt.Lifetime** - The total number of milliseconds that Cobalt has been
    running.

#### DebugCVals

*   **Cobalt.Server.DevTools** - The IP address and port of the DevTools server,
    if it is running.
*   **Cobalt.Server.WebDriver** - The IP address and port of the Webdriver
    server, if it is running.

### Count

#### PublicCVals

*   **Count.DOM.EventListeners** - The total number of EventListeners in
    existence globally. This includes ones that are pending garbage collection.
*   **Count.DOM.Nodes** - The total number of Nodes in existence globally. This
    includes ones that are pending garbage collection.
*   **Count.MainWebModule.DOM.HtmlElement** - The total number of HtmlElements
    in the MainWebModule. This includes elements that are not in the document
    and ones that are pending garbage collection.
*   **Count.MainWebModule.DOM.HtmlElement.Document** - The total number of
    HtmlElements that are in the MainWebModule’s document.
*   **Count.MainWebModule.DOM.HtmlScriptElement.Execute** - The total number of

>     HtmlScriptElement executions that have run since Cobalt started..

Reviewer: ```suggestion
    HtmlScriptElement executions that have run since Cobalt started.
```

*   **MainWebModule.Layout.IsRenderTreePending** - Whether or not the layout is
    scheduled to produce a new render tree.

### Media

#### PublicCVals

We have various SbPlayer metrics available which need to be enabled first using:
`h5vcc.settings.set("Media.EnableMetrics", 1)`

*   **Media.SbPlayerCreateTime.Minimum**
*   **Media.SbPlayerCreateTime.Median**
*   **Media.SbPlayerCreateTime.Maximum**
*   **Media.SbPlayerDestructionTime.Minimum**
*   **Media.SbPlayerDestructionTime.Median**
*   **Media.SbPlayerDestructionTime.Maximum**
*   **Media.SbPlayerWriteSamplesTime.Minimum**
*   **Media.SbPlayerWriteSamplesTime.Median**
*   **Media.SbPlayerWriteSamplesTime.Maximum**

#### DebugCVals

Media Pipeline exposes many State values, however we support multiple pipelines
at once, so we have to query for the current pipeline number to access them.
This can be done with `h5vcc.cVal.keys().filter(key =>
key.startsWith("Media.Pipeline.") && key.endsWith("MaxVideoCapabilities") &&
h5vcc.cVal.getValue(key).length === 0)` (If a Pipeline has MaxVideoCapabilities,
it implies that the Pipeline is a 10x player, i.e. a secondary small player, so

> if MaxVideoCapabilities.length is 0, then it must be the primary pipeline. )

Reviewer: ```suggestion
if MaxVideoCapabilities.length is 0, then it must be the primary pipeline.)
```
